RP235x support

CBJamo opened the pull request:
The primary purpose of this pr is to support the new rx fifo modes on the rp235x, it also includes the new JMP_PIN source for WAIT and PINDIRS destination for MOV.

I'm not happy that I had to make a second lalrpop file, but I couldn't figure out how to make conditional compilation work.

Also open to bike shedding on the feature name.

jannic commented:
> I'm not happy that I had to make a second lalrpop file, but I couldn't figure out how to make conditional compilation work.

Would it be possible to use a single lalrpop file if it always understood the additions of RP235x? If every valid RP2040 pio program is also valid on the RP235x, that should work, and as I understand it this is how the additions were designed.

CBJamo commented:
That's certainly possible, but then there wouldn't be compile time errors when trying to use a program with the new instructions on the 2040. Or at least there wouldn't be compile time errors originating in the lalrpop generated code like it would for any other invalid instruction.

jannic commented:
> That's certainly possible, but then there wouldn't be compile time errors when trying to use a program with the new instructions on the 2040. Or at least there wouldn't be compile time errors originating in the lalrpop generated code like it would for any other invalid instruction.

That's another advantage to do the parsing with rp235x support: There may be .pio files containing multiple programs, some rp2040 compatible and some with rp235x extensions. Such files should be usable on rp2040 as long as one only installs compatible programs. But with the current approach, the whole file would fail to parse.

Of course, now the question becomes at which point rp2040 compatibility should be checked. I did not yet think about it enough to have a clear picture.

One  possibility would be that `parse_file` / `parse_program` only return programs compatible with the current microcontroller. That way, when using `pio_proc`, one would still get a compile error, albeit with a more generic and maybe a bit misleading error message like "program name not found in the provided file".

We could also somehow note the PIO version inside the program and then check it in the macro (which could then provide a better error message). Not sure if that can be sensibly done without a breaking change, which would in turn necessitate a new version of rp2040-hal.

CBJamo commented:
Here is another attempt. I removed the rp2350 feature and added a version to the program struct, so pio-rs should be able to parse any program. I have mixed feelings about delegating handling the version to the HALs as I don't know how to make using an invalid program a compile time error there. This approach does make pio-rs much cleaner.

I also added support for the new Irq index modes. I think that was the only new feature (that belongs here in pio-rs) that was missing.

jannic commented:
@CBJamo: Sorry that I didn't come back to this PR earlier (and thanks to Dirbaio for reminding me on the matrix channel). Overall, I like the current version. Nice and clean. I agree that a version check at compile time would be nice, and in fact I wanted to look for a good solution for than - and then didn't have time for it, and forgot about the PR entirely :-(. Again: Sorry for that! Anyway, I don't think this should be a blocker for this PR. We can improve the HAL integration later.

I only had a quick look at the instruction encoding and didn't verify that the new things you added are correct. But the changes are quite readable and look like they don't break any existing usage. It'd be nice if you could restore the bounds check for the IRQ index, that's the only detail keeping me from approving this PR right now.

CBJamo commented on src/lib.rs:
```diff
@@ -259,20 +296,17 @@ impl InstructionOperands {
-            } => {
-                if *index > 7 {
-                    panic!("invalid interrupt flags");
-                }
```
I think I removed that in an earlier, different approach to the index_mode part of the IRQ instruction, I've added it back.

The MovToRx and MovFromRx instructions' index argument isn't just a number, it's an enum of the fifo locations. The DS just calls them "index", but I've renamed the struct members to fifo_index to make that a bit clearer.

jannic reviewed (approved):
LGTM, and thanks for the pull request and for reacting to all my comments so quickly!
